src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -132,23 +147,74 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
         /// </summary>
         private BoundDeconstructionDeconstructStep MakeNonTupleDeconstructStep(
                                                             BoundDeconstructValuePlaceholder targetPlaceholder,
-                                                            AssignmentExpressionSyntax syntax,
+                                                            ExpressionSyntax syntax,
                                                             DiagnosticBag diagnostics,
-                                                            ImmutableArray<DeconstructionVariable> variables,
+                                                            ArrayBuilder<DeconstructionVariable> variables,
                                                             ArrayBuilder<BoundDeconstructionDeconstructStep> deconstructionSteps,
                                                             ArrayBuilder<BoundDeconstructionAssignmentStep> assignmentSteps)
         {
             // symbol and parameters for Deconstruct
             ImmutableArray<BoundDeconstructValuePlaceholder> outPlaceholders;
-            var deconstructInvocation = MakeDeconstructInvocationExpression(variables.Length, targetPlaceholder, syntax, diagnostics, out outPlaceholders);
-
+            var deconstructInvocation = MakeDeconstructInvocationExpression(variables.Count, targetPlaceholder, syntax, diagnostics, out outPlaceholders);
             if (deconstructInvocation.HasAnyErrors)
             {
                 return null;
             }
-            else
+
+            SetInferredTypes(variables, outPlaceholders.SelectAsArray(p => p.Type));
+
+            return new BoundDeconstructionDeconstructStep(syntax, deconstructInvocation, targetPlaceholder, outPlaceholders);
+        }
+
+        /// <summary>
+        /// Inform the variables about found types (whether one was found or not).
+        /// </summary>
+        static private void SetInferredTypes(ArrayBuilder<DeconstructionVariable> variables, ImmutableArray<TypeSymbol> foundTypes)
+        {
+            if (!variables.Any(v => !v.IsNested && v.Single.Kind == BoundKind.DeconstructionLocalPendingInference))
+            {
+                return;
+            }
+
+            var varCount = variables.Count;
+            var foundCount = foundTypes.Length;
+
+            var matchCount = Math.Min(varCount, foundCount);
+            for (int i = 0; i < varCount; i++)
+            {
+                var variable = variables[i];
+                if (!variable.IsNested && variable.Single.Kind == BoundKind.DeconstructionLocalPendingInference)
```

Comment:
This same expression is used in the lambda above. Consider extracting a helper method.